aspnetcore/blazor/security/server/qrcodes-for-authenticator-apps.md

```diff
@@ -0,0 +1,224 @@
+
+1. In **Solution Explorer**, right-click **Add** > **New Scaffolded Item**.
+1. Select **Identity** > **Blazor Identity**. Select the **Add** button.
+1. On the **Add Blazor Identity** step, override the file for **Pages\Manage\EnableAuthenticator** in the list of pages. Select the app's DBContext class. Select the **Add** button.
+
+<!-- UPDATE 9.0 Update NOTE per followup on the issue -->
+
+> [!NOTE]
+> Although only the `EnableAuthenticator` component is selected for scaffolding in this example, scaffolding currently adds all of the Identity components to the app. Additionally, exceptions may be thrown during the process of scaffolding into the app. If exceptions occur when database migrations occur, stop the app and restart it on each exception. For more information, see [Scaffolding exceptions for Blazor Web App (`dotnet/Scaffolding` #2694)](https://github.com/dotnet/Scaffolding/issues/2694).
+
+Be patient while migrations are executed. Depending on the speed of the system, it can take up to a minute or two for database migrations to finish.
+
+For more information, see <xref:security/authentication/scaffold-identity>. For guidance on using the .NET CLI instead of Visual Studio, see <xref:fundamentals/tools/dotnet-aspnet-codegenerator>.
+
+## Adding QR codes to the 2FA configuration page
+
+These instructions use [Shim Sangmin](https://hogangnono.com)'s [qrcode.js: Cross-browser QRCode generator for JavaScript](https://davidshimjs.github.io/qrcodejs/) ([`davidshimjs/qrcodejs` GitHub repository](https://github.com/davidshimjs/qrcodejs)).
+
+Download the [`qrcode.min.js`](https://davidshimjs.github.io/qrcodejs/) library to the `wwwroot` folder of the solution's server project. The library has no dependencies.
+
+In the `App` component (`Components/App.razor`), place a library script reference after [Blazor's `<script>` tag](xref:blazor/project-structure#location-of-the-blazor-script):
+
+```razor
+<script src="qrcode.min.js"></script>
+```
+
+The `EnableAuthenticator` component, which is part of the QR code system in the app and displays the QR code to users, adopts static server-side rendering (static SSR) with enhanced navigation. Therefore, normal scripts can't execute when the component loads or updates under enhanced navigation. Extra steps are required to trigger the QR code to load in the UI when the page is loaded. To accomplish loading the QR code, the approach explained in <xref:blazor/js-interop/ssr> is adopted.
+
+Add the following [JavaScript initializer](xref:blazor/fundamentals/startup#javascript-initializers) to the server project's `wwwroot` folder. The `{NAME}` placeholder must be the name of the app's assembly in order for Blazor to locate and load the file automatically. If the server app's assembly name is `BlazorSample`, the file is named `BlazorSample.lib.module.js`.
+
+`wwwroot/{NAME}.lib.module.js`:
+
+```javascript
+const pageScriptInfoBySrc = new Map();
```

Comment:
To keep this section shorter, would we consider the one of the following alternatives?

1. Link to the article about setting up `PageScript` without including all that content here, or
2. Use an approach similar to the one you sent me via email, but rather than using the "load" event, use Blazor's `enhancedload` event, which runs after every enhanced navigation:
    ```js
    Blazor.addEventListener('enhancedload', () => { /* Generate the QR code if the target element exists */ });
    ```

I would probably prefer option [2] if possible because it requires less machinery than going all-out with the `<PageScript>` component.

By the way - I created https://github.com/MackinnonBuck/blazor-page-script as a convenient way to add `<PageScript>` to an app so that customers don't have to copy and paste code. I'm not sure if we ever point to non-Microsoft projects from the docs, but if we do, it might be worth mentioning that it's an easier alternative to copying all the code into your project (at the cost of not being able to customize its implementation).

Reply:
I'll refactor for `enhancedload`. I like that, too, because the JS node article doesn't have a working example of that approach.

On linking the external repo ... yes ... we do take that approach sometimes. I'm not sure about doing it in the JS article. I'd leave it up to you, Dan, and Artak to decide if that's what you guys want. Also, I'd work that as a separate lower-priority issue. I'm still swamped with higher priority work, and that's not super pressing at the moment, right?

Reply:
BTW ... I'm not 100% clear on the `blazor.addEventListener` approach. Where does that go? ... not collocated in a module? If not, that's why you're saying to check if the element exists? Don't devs consider that a code smell :poop: to check constantly for the existence for an element, when the element only appears on one or a few pages?

Reply:
It could go either in `afterWebStarted` or a `<script>` element just after the one for `blazor.web.js`. Yeah, it could be considered a code smell for the logic to run more often than it needs to, but the alternative is a more sophisticated approach like `PageScript` that uses additional logic to only activate logic that's specific to the current page. I suppose it's somewhat of a balance between having a simpler example vs one that cleanly separates different concerns.

Reply:
Well ... if it were me, I'd go with the `PageScript` approach in my own stuff. Are you sure I shouldn't just link over to that article, dropping the `PageScript` here. After all, if they're cool with the `addEventListener` approach, they can go that way based on the content of the JS interop article (although I admit we don't have an actual fully working example there).

Reply:
> Are you sure I shouldn't just link over to that article (the JS interop one), dropping the PageScript here (from the QR code article).

Sure, that sounds fine to me! If you think it would help, maybe it would be worth mentioning (in just a sentence, maybe) that using `Blazor.addEventListener('enhancedload', ...)` is a viable alternative approach.

Reply:
Very well ... I'll do that and mention it in passing here.

Reply:
I was able to add the language on `Blazor.addEventListener('enhancedload', ...)` easily enough, but bisecting this guidance doesn't work well. I don't have a clean way to only refer readers to the JS initializer code in the middle of a sequence of steps that pertains to adding it to an RCL from a sequence of steps that focuses on adding it directly to the BWA. It would also be unnatural to introduce artificial sections to the JS interop article merely for this purpose. I'm not favorable to cross-linking to a separate code file that would open in a new browser window either. It's just not good user experience either way. The best thing for readers is just to get the code they need when they need it working straight down the steps.

However, I'll place the code into an INCLUDE file and link it from there. That way, we can change the code in one spot and update article dates to draw it into both articles.